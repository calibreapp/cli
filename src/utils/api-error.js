const handleError = err => {
  if (err.response && err.response.error) throw err.response
  if (err.response && err.response.errors) throw err.response.errors
  else throw err
}

const humaniseError = apiError => {
  if (
    apiError[0] &&
    apiError[0].extensions &&
    apiError[0].extensions.problems &&
    apiError[0].extensions.problems[0].explanation
  ) {
    return apiError[0].extensions.problems[0].explanation
  } else if (apiError[0] && apiError[0].message) {
    return apiError[0].message
  } else if (apiError.message) {
    return apiError.message
  } else {
    return `An unknown error occurred ${JSON.stringify(apiError, null, 2)}`
  }
}

module.exports = {
  handleError,
  humaniseError
}
