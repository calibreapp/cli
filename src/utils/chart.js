const chalk = require('chalk')
const { format } = require('../utils/formatters')

const barSection = 'â– '
const chartWidth = 80

const bar = (value, maxValue) => {
  const barLength = (value * chartWidth) / maxValue
  const wholeNumberPart = Math.floor(barLength)
  const emptyPart = chartWidth - wholeNumberPart
  let bar = chalk.blue(barSection).repeat(wholeNumberPart)
  let empty = chalk.grey(barSection).repeat(emptyPart)

  return bar + empty
}

const chart = (data, formatter) => {
  const maxValue = Math.max(...data.map(item => parseInt(item.value || 0)))
  return data
    .map(item => {
      const barText = bar(item.value || 0, maxValue)
      return `${item.key}\n${barText} ${format({
        formatter,
        value: item.value
      })}`
    })
    .join('\n\n')
}

module.exports = chart
