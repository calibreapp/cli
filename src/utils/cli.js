const options = {
  site: {
    demandOption: true,
    describe: 'The identifying slug of a site',
    type: 'string',
    requiresArg: true
  },
  json: { describe: 'Return the output in JSON format' },
  csv: { describe: 'Return the output in CSV format' },
  from: {
    describe:
      'The start date to retrieve from in the format Year-Month-Day (when not set, defaults to 7 days ago)'
  },
  to: {
    describe:
      'The end date to retrieve to in the format Year-Month-Day (when not set, defaults to today)'
  },
  count: {
    default: 25,
    describe: 'The number of records to return, maximum: 500'
  },
  cursor: {
    describe: 'The cursor to fetch records after'
  }
}

module.exports.options = options
