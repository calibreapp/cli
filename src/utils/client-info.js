import pkg from '../../package.json'

const name = 'calibre-node'
const version = pkg.version

export { name, version }
