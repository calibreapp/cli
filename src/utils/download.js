const https = require('https')
const zlib = require('zlib')
const fs = require('fs')

const download = (url, destination) => {
  if (fs.existsSync(destination)) return Promise.resolve()

  return new Promise(resolve => {
    const file = fs.createWriteStream(destination)
    https.get(url, response => {
      if (response.headers['content-encoding'] == 'gzip') {
        response.pipe(zlib.createGunzip()).pipe(file)
      } else {
        response.pipe(file)
      }

      file.on('finish', () => file.close(resolve))
    })
  })
}

module.exports = download
