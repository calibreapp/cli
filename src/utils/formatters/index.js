const filesize = require('./filesize')
const duration = require('./duration')
const gradeFormatter = require('./grade-score')
const milliunit = require('./milliunit')

const format = ({ formatter, value }) => {
  switch (formatter) {
    case 'humanDuration':
      return duration(value)
    case 'fileSize':
      return filesize(value)
    case 'milliunit':
      return milliunit(value)
    default:
      return value
  }
}

module.exports = {
  format,
  filesize,
  duration,
  gradeFormatter,
  milliunit
}
