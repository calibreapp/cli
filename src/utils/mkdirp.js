const fs = require('fs')
const path = require('path')

// Returns a string of the created directory path
/*
 * @param {string[]} - The paths to create
 * @return {string} - The full path to the resulted directory
 */
const mkdirp = directories => {
  directories.reduce((reducePath, dir) => {
    reducePath = path.join(reducePath, dir)
    if (!fs.existsSync(reducePath)) {
      fs.mkdirSync(reducePath)
    }
    return reducePath
  }, '')

  return path.join(...directories)
}

module.exports = mkdirp
