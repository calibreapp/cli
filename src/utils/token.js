const Configstore = require('configstore')
const pkg = require('../../package.json')
const config = new Configstore(pkg.name)

const retrieveToken = () => {
  const token = process.env.CALIBRE_API_TOKEN || config.get('token')
  if (!token) {
    throw new Error(
      'Please set your Calibre API token by running `calibre token set`\n or set the environment variable CALIBRE_API_TOKEN.\n\n See calibreapp.com/docs/api/tokens for help.'
    )
  }

  return token
}

module.exports = retrieveToken
