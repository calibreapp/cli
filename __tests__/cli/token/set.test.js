import { runCLI } from '../../utils'

test('missing token argument', async () => {
  const out = await runCLI({ args: 'token', testForError: true })
  expect(out).toMatchSnapshot()
})

test('sets token', async () => {
  const out = await runCLI({ args: 'token set 123', testForError: true })
  expect(out).toMatchSnapshot()
})
