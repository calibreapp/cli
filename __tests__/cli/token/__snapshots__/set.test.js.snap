// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`missing token argument 1`] = `
"

 Not enough non-option arguments: got 0, need at least 1 

 calibre token <command>

Use the token command to store or revoke a Calibre API token. You can also save
your token with the CALIBRE_API_TOKEN environment variable.

Commands:

  calibre token set <token>  Store your API Token to use the CLI (saved in ~/.co
                             nfig/configstore/calibre.json).
  calibre token remove       Remove the saved API Token used for CLI commands (f
                             rom ~/.config/configstore/calibre.json).

Options:

  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]"
`;

exports[`sets token 1`] = `""`;
